flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.hadoop.mapreduce.utils.HadoopUtils;
+import org.apache.flink.api.java.typeutils.GenericTypeInfo;
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.WritableComparable;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.JobConfigurable;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.OutputCommitter;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hive.hcatalog.common.HCatUtil;
+import org.apache.hive.hcatalog.data.HCatRecord;
+import org.apache.hive.hcatalog.data.schema.HCatFieldSchema;
+import org.apache.hive.hcatalog.data.schema.HCatSchema;
+import org.apache.hive.hcatalog.mapreduce.OutputJobInfo;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.List;
+import java.util.Map;
+
+public abstract class HCatOutputFormatBase<T> implements OutputFormat<T>, FinalizeOnMaster {
+	private static final long serialVersionUID = 1L;
+
+	private Configuration configuration;
+	//a Job is used to create OutputJobInfo
+	private Job job;
+
+	private org.apache.hive.hcatalog.mapreduce.HCatOutputFormat hCatOutputFormat;
+	private RecordWriter<WritableComparable<?>, HCatRecord> recordWriter;
+
+	//required data type by the table
+	protected TypeInformation<T> reqType;
+
+	protected String[] fieldNames = new String[0];
+	protected HCatSchema outputSchema;
+	private transient TaskAttemptContext context;
+
+
+	@Override
+	public void configure(org.apache.flink.configuration.Configuration parameters) {
+		// configure MR OutputFormat if necessary
+		if (this.hCatOutputFormat instanceof Configurable) {
+			((Configurable) this.hCatOutputFormat).setConf(configuration);
+		} else if (this.hCatOutputFormat instanceof JobConfigurable) {
+			((JobConfigurable) this.hCatOutputFormat).configure(new JobConf(configuration));
+		}
+	}
+
+	/**
+	 * Creates a HCatOutputFormat for the given database and table and a Map of PartitionValues
+	 * By default. The RecordWriter of
+	 * {@link HCatInputFormatBase#asFlinkTuples()}.
+	 *
+	 * @param database The name of the database to read from.
+	 * @param table    The name of the table to read.
+	 * @param partitionValues the map of partition values. if null, the job writes to unpartitioned
+	 *                        table
+	 * @throws java.io.IOException
+	 */
+	public HCatOutputFormatBase(String database, String table, Map<String, String> partitionValues)
+			throws IOException {
+		this(database, table, partitionValues, new Configuration());
+	}
+
+	protected abstract HCatRecord TupleToHCatRecord(T record) throws IOException;
```

Comment:
Other name such as `convertTupleToHCatRecord` would be better. The name of method should start with lower case.